linera-execution/src/evm/revm.rs

```diff
@@ -728,17 +762,33 @@ impl<Runtime: ContractRuntime> CallInterceptorContract<Runtime> {
 
 struct CallInterceptorService<Runtime> {
     db: DatabaseRuntime<Runtime>,
+    // This is the caller contract address
+    contract_address: Address,
+    precompile_addresses: BTreeSet<Address>,
 }
 
 impl<Runtime> Clone for CallInterceptorService<Runtime> {
     fn clone(&self) -> Self {
         Self {
             db: self.db.clone(),
+            contract_address: self.contract_address,
+            precompile_addresses: self.precompile_addresses.clone(),
         }
     }
 }
 
 impl<'a, Runtime: ServiceRuntime> Inspector<Ctx<'a, Runtime>> for CallInterceptorService<Runtime> {
+    fn create(
```

Comment:
This `impl` has the same body for `ServiceRuntime` and `ContractRuntime`. Can't it be shared via an impl for `Runtime: BaseRuntime` (i.e. a Runtime that implements `BaseRuntime` - a common trait for both `Service*` and `Contract*` runtimes.